tools/draincluster/drain/drain.go

```diff
@@ -0,0 +1,244 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package drain
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	k8errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	clusterv1beta1 "go.goms.io/fleet/apis/cluster/v1beta1"
+	placementv1beta1 "go.goms.io/fleet/apis/placement/v1beta1"
+	"go.goms.io/fleet/pkg/utils"
+	"go.goms.io/fleet/pkg/utils/condition"
+	evictionutils "go.goms.io/fleet/pkg/utils/eviction"
+	toolsutils "go.goms.io/fleet/tools/utils"
+)
+
+const (
+	drainEvictionNameFormat     = "drain-eviction-%s-%s"
+	resourceIdentifierKeyFormat = "%s/%s/%s/%s/%s"
+)
+
+type Helper struct {
+	HubClient                            client.Client
+	ClusterName                          string
+	ClusterResourcePlacementResourcesMap map[string][]placementv1beta1.ResourceIdentifier
+}
+
+func (h *Helper) Drain(ctx context.Context) (bool, error) {
+	if err := h.cordon(ctx); err != nil {
+		return false, fmt.Errorf("failed to cordon member cluster %s: %w", h.ClusterName, err)
+	}
+
+	if err := h.fetchClusterResourcePlacementToEvict(ctx); err != nil {
+		return false, err
+	}
+
+	if len(h.ClusterResourcePlacementResourcesMap) == 0 {
+		log.Printf("There are currently no resources propagated to %s from fleet using ClusterResourcePlacement resources", h.ClusterName)
+		return true, nil
+	}
+
+	// create eviction objects for all <crpName, targetCluster>.
+	for crpName := range h.ClusterResourcePlacementResourcesMap {
+		evictionName := fmt.Sprintf(drainEvictionNameFormat, crpName, h.ClusterName)
+
+		if err := removeExistingEviction(ctx, h.HubClient, evictionName); err != nil {
```

Comment:
I have some reservations in this part; eviction objects are left in the cluster after its creation/execution as a track record, now it is being erased every time the tool is called.

Reply:
I understand that this part is needed for conflict resolution, but would it be better if we just use a random suffix in name? (considering that the eviction controller can handle this situation)